src/System.Text.Json/src/System/Text/Json/Writer/Utf8JsonWriter.WriteProperties.UnsignedNumber.cs

```diff
@@ -0,0 +1,312 @@
+
+            WriteNumberByOptions(ref propertyName, value);
+
+            if (propertyArray != null)
+                ArrayPool<byte>.Shared.Return(propertyArray);
+        }
+
+        private void WriteNumberByOptions(ref ReadOnlySpan<char> propertyName, ulong value)
+        {
+            if (_writerOptions.Indented)
+            {
+                if (!_writerOptions.SkipValidation)
+                {
+                    ValidateWritingProperty();
+                }
+                WriteNumberIndented(ref propertyName, value);
+            }
+            else
+            {
+                if (!_writerOptions.SkipValidation)
+                {
+                    ValidateWritingProperty();
+                }
+                WriteNumberMinimized(ref propertyName, value);
+            }
+        }
+
+        private void WriteNumberByOptions(ref ReadOnlySpan<byte> propertyName, ulong value)
+        {
+            if (_writerOptions.Indented)
+            {
+                if (!_writerOptions.SkipValidation)
+                {
+                    ValidateWritingProperty();
+                }
+                WriteNumberIndented(ref propertyName, value);
+            }
+            else
+            {
+                if (!_writerOptions.SkipValidation)
+                {
+                    ValidateWritingProperty();
+                }
+                WriteNumberMinimized(ref propertyName, value);
+            }
+        }
+
+        private void WriteNumberMinimized(ref ReadOnlySpan<char> escapedPropertyName, ulong value)
+        {
+            int idx = WritePropertyNameMinimized(ref escapedPropertyName);
+
+            WriteNumberValueFormatLoop(value, ref idx);
+
+            Advance(idx);
+        }
+
+        private void WriteNumberMinimized(ref ReadOnlySpan<byte> escapedPropertyName, ulong value)
+        {
+            int idx = WritePropertyNameMinimized(ref escapedPropertyName);
+
+            WriteNumberValueFormatLoop(value, ref idx);
+
+            Advance(idx);
+        }
+
+        private void WriteNumberIndented(ref ReadOnlySpan<char> escapedPropertyName, ulong value)
+        {
+            int idx = WritePropertyNameIndented(ref escapedPropertyName);
+
+            WriteNumberValueFormatLoop(value, ref idx);
+
+            Advance(idx);
+        }
+
+        private void WriteNumberIndented(ref ReadOnlySpan<byte> escapedPropertyName, ulong value)
+        {
+            int idx = WritePropertyNameIndented(ref escapedPropertyName);
+
+            WriteNumberValueFormatLoop(value, ref idx);
+
+            Advance(idx);
+        }
+
+        private void WriteNumberValueFormatLoop(ulong value, ref int idx)
+        {
+            int bytesWritten;
+            // Using Utf8Formatter with default StandardFormat is roughly 30% slower (17 ns versus 12 ns)
+            // See: https://github.com/dotnet/corefx/issues/25425
+            // Utf8Formatter.TryFormat(value, _buffer.Slice(idx), out bytesWritten);
```

Comment:
Then we should fix Utf8Formatter, not duplicate the code.

Reply:
I have brought this issue up with @AndyAyersMS, offline already, but I agree. I will remove this workaround while we try to optimize the `Utf8Formatter` to avoid such code duplication. I didn't want to do it to begin with :)

> does it only impact a microbenchmark on writing numbers, or does this actually show up as a meaningful impact on the whole scenario?

It depends on what you consider meaningful. I haven't looked at larger end-to-end scenario, but writing a general JSON payload with some number values slows down by ~10%.

```C#
json.WriteStartObject();
json.WriteNumber("age", 42);
json.WriteString("first", "John");
json.WriteString("last", "Smith");
json.WriteStartArray("phoneNumbers");
json.WriteStringValue("425-000-1212");
json.WriteStringValue("425-000-1213");
json.WriteEndArray();
json.WriteStartObject("address");
json.WriteString("street", "1 Microsoft Way");
json.WriteString("city", "Redmond");
json.WriteNumber("zip", 98052);
json.WriteEndObject();

json.WriteStartArray("ExtraArray");
for (var i = 0; i < data.Length; i++)
{
    json.WriteNumberValue(data[i]);
}
json.WriteEndArray();

json.WriteEndObject();

json.Flush();
```

``` ini

BenchmarkDotNet=v0.11.3, OS=Windows 10.0.17763.195 (1809/October2018Update/Redstone5)
Intel Core i7-6700 CPU 3.40GHz (Skylake), 1 CPU, 8 logical and 4 physical cores
.NET Core SDK=3.0.100-preview-010006
  [Host] : .NET Core 3.0.0-preview-27218-01 (CoreCLR 4.6.27217.02, CoreFX 4.7.18.61304), 64bit RyuJIT
  Core   : .NET Core 3.0.0-preview-27218-01 (CoreCLR 4.6.27217.02, CoreFX 4.7.18.61304), 64bit RyuJIT

Job=Core  EnvironmentVariables=COMPlus_TieredCompilation=0  Runtime=Core  
Toolchain=.NET Core 3.0  IterationCount=5  WarmupCount=3  

```
|             Method | Formatted | SkipValidation |       Mean |    Error |   StdDev | Ratio |
|------------------- |---------- |--------------- |-----------:|---------:|---------:|------:|
|     SystemTextJson |     False |           True |   935.6 ns | 19.88 ns | 5.164 ns |  1.00 |
| SystemTextJson_new |     False |           True | 1,007.4 ns | 14.55 ns | 3.780 ns |  1.08 |


Whether this regression matters a lot, is another question.